### ISSUE TITLE
Using non-global class for plugin custom type throws an error

### ISSUE THREAD

**Issue by adamscott** [godotengine/godot#69889]
### Godot version

4.0.dev (a4131b6)

### System information

Linux (Ubuntu 22.04), Vulkan, Radeon RX Vega 64 (eGPU)

### Issue description

When trying to add a custom node from a plugin that don't come from a global class, an error shows up in the editor when the Add Node window is displayed.

```
ERROR: Cannot get class 'ExtendedNode'.
   at: is_virtual (core/object/class_db.cpp:382)
```

Note that `ExtendedNode` is registered as a custom node, but the `GDScript` instance used comes from a constant, not a global class.

```gdscript
# plugin.gd
@tool
extends EditorPlugin

const Icon: = preload("res://icon.svg")
const ExtendedNode: = preload("extended_node.gd")


func _enter_tree() -> void:
	add_custom_type("ExtendedNode", "Node", ExtendedNode, Icon)

func _exit_tree() -> void:
	remove_custom_type("ExtendedNode")
```

The error thrown seems to exist since #67614, but it seems rather linked to the `ClassDB::is_virtual()` addition (#58972). Despite the error, the node seems to work well.

It should be possible to use non-global classes to create new types of nodes.

### Steps to reproduce

1. Open the project
2. Click on "Add Node" on the top left
3. See that the error is thrown

### Minimal reproduction project

[non_global_plugin.zip](https://github.com/godotengine/godot/files/10201314/non_global_plugin.zip)

**Comment by adamscott** [comment]
CC @aaronfranke

**Comment by aaronfranke** [comment]
CC @willnationsdev, if I recall correctly the plan is to fully replace the `add_custom_type` system with global classes in Godot 4? 

> It should be possible to use non-global classes to create new types of nodes.

Should it? What prevents you from using `class_name` for your use cases here?

**Comment by KoBeWi** [comment]
Why not just
```GDScript
add_custom_type("ExtendedNode", "Node", preload("extended_node.gd"), preload("res://icon.svg"))
```
?

**Comment by adamscott** [comment]
> Why not just
> 
> ```gdscript
> add_custom_type("ExtendedNode", "Node", preload("extended_node.gd"), preload("res://icon.svg"))
> ```

Yes, maybe, but it doesn't change the outcome. The same error pops up.

**Comment by willnationsdev** [comment]
`add_custon_type()` is nothing more than window-dressing / an editor-only macro for creating a node and attaching a script to it along with a default name and a user-specified icon. It just makes it a new option for a kind of node to create from the CreateDialog.

It doesn't *actually* register a class type by any means. There is no static typing support, no exporting of resources, etc. It is a legacy tool that should be well and truly deprecated and removed for 4.0 imo, but I know that some people have some use cases for them (though what those use cases are is not something with which I am really familiar).

As mentioned, some people have insisted on maintaining it (for whatever reasons). But provided we can update global classes to satisfy the use case(s), there is really no reason that Godot 4 should continue to support it at all: it is bound to be removed at *some* point in the near future regardless. It is misleading for people unfamiliar with it. It also adds to the complexity of the codebase despite the fact that it doesn't provide nearly the same depth of functionality.

**Comment by aaronfranke** [comment]
The reason this bug exists in the first place is fairly simple.

I added a check for whether or not the class is virtual to the Create New Node dialog, so that we can gray them out. This check assumes that everything in the dialog is a global class. `add_custom_type` breaks that assumption.

So yeah, my point is that the best solution is likely to just remove `add_custom_type`. After which, all custom types would be created with the `class_name` system, and everything in the Create New Node dialog would be a global class.

**Comment by adamscott** [comment]
> So yeah, my point is that the best solution is likely to just remove `add_custom_type`. After which, all custom types would be created with the `class_name` system, and everything in the Create New Node dialog would be a global class.

@aaronfranke Seems that the global class solution has problems with C#: #69534 

CC @neikeq @raulsntos

**Comment by raulsntos** [comment]
Global classes are currently not supported in C# (see https://github.com/godotengine/godot/pull/63126 that adds support for global classes in C# when used as a scripting language), so C# developers currently use `add_custom_type` as a workaround.

Alternatively, GDExtension can also be used as a replacement of `add_custom_type` since it would allow registering types in ClassDB as if they were built-in classes, but it's also not currently supported in C# yet.

I'm not sure I understand the issue linked (https://github.com/godotengine/godot/issues/69534), there doesn't seem to be a C# MRP. If they are just making an EditorPlugin that registers a C# class as a custom type, that used to work in 3.x but I'm not sure about 4.x.

**Comment by adamscott** [comment]
> So yeah, my point is that the best solution is likely to just remove `add_custom_type`.

I don't think this is feasible for the 4.0 release, unless we consider a fix in C# as needed for the release.

**Comment by raulsntos** [comment]
> I added a check for whether or not the class is virtual to the Create New Node dialog, so that we can gray them out. This check assumes that everything in the dialog is a global class

The check only seems to execute if the class exists in ClassDB, so custom types should be unaffected.

https://github.com/godotengine/godot/blob/05097ded0a915cd6c083f15dab08da2bdc0770b8/editor/create_dialog.cpp#L129-L130

The error actually comes from `_configure_search_option_item` because the `ClassDB::is_virtual` method is invoked without checking if the type is in ClassDB first, so wouldn't adding that check be enough?

> > So yeah, my point is that the best solution is likely to just remove add_custom_type.
>
> I don't think this is feasible for the 4.0 release, unless we consider a fix in C# as needed for the release.

Removing `add_custom_type` would be a breaking change, and we may be too late for that. I'd be in favor of removing it if we get Global classes and/or GDExtension support in C# (which seems like it will happen for 4.0 but I don't want to make any promises).

As long as we don't have a replacement for `add_custom_type` in C#, removing it would be kind of a big regression coming from 3.x and I'm sure a lot of users won't like it. If we can't or don't want to fix `add_custom_type` maybe we can consider leaving it in 4.0 as deprecated and remove it in 4.1 when we have a good alternative for users.
